roofit/roofit/src/RooPolyFunc.cxx

```diff
@@ -0,0 +1,287 @@
+
+RooPolyFunc::RooPolyFunc(const char *name, const char *title, const RooAbsCollection &vars)
+   : RooAbsReal(name, title), _vars("x", "list of dependent variables", this)
+{
+   for (const auto &var : vars) {
+      if (!dynamic_cast<RooAbsReal *>(var)) {
+         std::stringstream ss;
+         ss << "RooPolyFunc::ctor(" << GetName() << ") ERROR: coefficient " << var->GetName()
+            << " is not of type RooAbsReal";
+         const std::string errorMsg = ss.str();
+         coutE(InputArguments) << errorMsg << std::endl;
+         throw std::runtime_error(errorMsg);
+      }
+      _vars.add(*var);
+   }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Copy constructor
+
+RooPolyFunc::RooPolyFunc(const RooPolyFunc &other, const char *name)
+   : RooAbsReal(other, name), _vars("vars", this, other._vars)
+{
+   for (size_t i = 0; i < other._terms.size(); ++i) {
+      this->_terms.push_back(std::make_unique<RooListProxy>(other._terms[i]->GetName(), this, *(other._terms[i])));
+   }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Evaluate value of Polynomial.
+double RooPolyFunc::evaluate() const
+{
+   // Calculate and return value of polynomial
+   double poly_sum(0.0);
+   for (const auto &term : _terms) {
+      double poly_term(1.0);
+      size_t n_vars = term->size() - 1;
+      for (size_t i_var = 0; i_var < n_vars; ++i_var) {
+         auto var = dynamic_cast<RooRealVar *>(_vars.at(i_var));
+         auto exp = dynamic_cast<RooRealVar *>(term->at(i_var));
+         poly_term *= pow(var->getVal(), exp->getVal());
+      }
+      auto coef = dynamic_cast<RooRealVar *>(term->at(n_vars));
+      poly_sum += coef->getVal() * poly_term;
+   }
+   return poly_sum;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Compute multiple values of Polynomial.
+// RooSpan<double> RooPolyFunc::evaluateSpan(RooBatchCompute::RunContext& evalData, const RooArgSet* normSet) const {
+// }
+
+////////////////////////////////////////////////////////////////////////////////
+/// Taylor expanding given function in terms of observables around
+/// observableValues.
+
+std::unique_ptr<RooAbsReal> RooPolyFunc::taylorExpand(RooAbsReal &func, int order, const RooAbsCollection &observables,
+                                                      std::vector<double> const &observableValues)
+{
+   // create the taylor expansion polynomial
+   auto te_name = std::string("taylorExpand_") + func.GetName() + std::string("_order_") + std::to_string(order);
+   auto taylor_poly = std::make_unique<RooPolyFunc>(te_name.c_str(), te_name.c_str(), observables);
+
+   // taylor expansion can be performed only for order 0, 1, 2 currently
+   if (order >= 3 || order <= 0) {
+      // oocoutE(taylor_poly,InputArguments) << "RooPolyFunc::ctor(" << te_name.c_str() <<
+      // ") ERROR: order must be 0, 1, or 2" << endl;
+      return NULL;
```

Comment:
```suggestion
      std::stringstream errorMsgStream;
      errorMsgStream << "RooPolyFunc::taylorExpand(" << te_name << ") ERROR: order must be 0, 1, or 2";
      const auto errorMsg = errorMsgStream.str();
      oocoutE(taylor_poly.get(), InputArguments) << errorMsg << std::endl;
      throw std::invalid_argument(errorMsg);
```
You will get obscure segmentation faults if you handle this error by returning a `nullptr`. It would be better to throw an exception here. Also make sure that `<sstream>` and `<exception>` are included in this file.